### PR TITLE
Fix #282 - Remove allocating lambda in GetService

### LINKED ISSUE #282: [PERF] Remove lambda allocation in ServiceProvider.GetService
https://github.com/aspnet/DependencyInjection/blob/dev/src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs#L52

This lamdba is allocating about 6mb across 10000 requests of a simple MVC app. We could easily make this 0 by turning the lambda into an instance method and caching a delegate to it.

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs [hunk 1]
```diff
@@ -49,17 +56,19 @@ internal ServiceProvider(ServiceProvider parent)
...
+            if (callSite != null)
             {
-                var callSite = GetServiceCallSite(key, new HashSet<Type>());
-                if (callSite != null)
-                {
-                    return RealizeService(_table, key, callSite);
-                }
-                return _ => null;
-            });
+                return RealizeService(_table, serviceType, callSite);
```
Comment (davidfowl) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: Doesn't this allocate too?

Comment (rynowak) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: Yes, but only once, then it's cached by `ServiceTable`. The problem described in the bug is one endemic to `GetOrAdd` - you usually have to allocate just to call the function.

**Comment by pranavkm** [comment]
:shipit:

File: src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: I see

**Pull request closed without merging**